### PR TITLE
JSON API

### PR DESCRIPTION
**Pull request by gr2m**
closes #5

I've cleaned up the commits, but left the old PR (#5) in tact
for simpler review

@patriciagarcia

### CODE REVIEW

**Comment by garrensmith** [comment]
This looks really really nice. Much clearer and I can understand now why you want this as a hapi api plugin.

**Comment by gr2m** [comment]
Thanks for the review @garrensmith. Glad to you the direction this is heading :)

File: api/README.md [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+var api = require ('hapi-couchdb-account-api/api')({
+  url: couchdbUrl
+})
+```
+
+All methods return promises
+
+```js
+api.sessions.add({username: 'pat', password: 'secret'/*, [jsonApiOptions] */})
+api.sessions.find(id /*, jsonApiOptions */)
```
Comment (patriciagarcia) [inline api/README.md]: `id` here is the `id:xxxx` role or `username`?

File: api/accounts/add.js [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+
+var createAccount = require('../../utils/account/create')
+
+function addSession (state, options) {
+  return new Promise(function (resolve, reject) {
+    createAccount({
+      couchUrl: state.url,
+      username: options.username,
+      password: options.password,
+      includeProfile: options.include === 'account.profile'
```
Comment (patriciagarcia) [inline api/accounts/add.js]: do you need to mention in [here](http://docs.accountrestapi.apiary.io/#reference/user/session/sign-out) that `?include=account.profile` is supported?

File: api/accounts/find.js [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+      username: username,
+      includeProfile: options.include === 'profile'
+    }, function (error, doc) {
+      if (error) {
+        return reject(error)
+      }
+
+      var account = {
+        username: doc.name,
+        id: doc.roles[0].substr(3)
```
Comment (patriciagarcia) [inline api/accounts/find.js]: can you be sure that id role is the first element of the array?

File: api/accounts/remove.js [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+
+var deleteAccount = require('../../utils/account/delete')
+
+function removeAccount (state, username, options) {
+  return new Promise(function (resolve, reject) {
+    deleteAccount({
+      couchUrl: state.url,
+      username: username,
+      bearerToken: options.bearerToken,
+      includeProfile: options.include === 'account.profile'
```
Comment (patriciagarcia) [inline api/accounts/remove.js]: do you need to mention in [here](http://docs.accountrestapi.apiary.io) that `?include=account.profile` is supported?

File: api/accounts/add.js [hunk 5]
```diff
@@ -0,0 +1,20 @@
+module.exports = addSession
```
Comment (patriciagarcia) [inline api/accounts/add.js]: `addAccount` will be a better name for this :-)

File: api/sessions/add.js [hunk 6]
```diff
@@ -0,0 +1,20 @@
...
+
+var createSession = require('../../utils/session/create')
+
+function addSession (state, options) {
+  return new Promise(function (resolve, reject) {
+    createSession({
+      couchUrl: state.url,
+      username: options.username,
+      password: options.password,
+      includeProfile: options.include === 'account.profile'
```
Comment (patriciagarcia) [inline api/sessions/add.js]: do we need to mention in [here](http://docs.accountrestapi.apiary.io) that `?include=account.profile` is supported?

File: api/sessions/remove.js [hunk 7]
```diff
@@ -0,0 +1,20 @@
...
+
+var deleteSession = require('../../utils/session/delete')
+
+function removeSession (state, id, options) {
+  return new Promise(function (resolve, reject) {
+    deleteSession({
+      couchUrl: state.url,
+      bearerToken: id,
+      includeAccount: options.include === 'account',
+      includeProfile: options.include === 'account.profile'
```
Comment (patriciagarcia) [inline api/sessions/remove.js]: do we need to mention in [here](http://docs.accountrestapi.apiary.io) that `?include=account` and `?include=account.profile` are supported?

File: api/README.md [hunk 8]
```diff
@@ -0,0 +1,34 @@
...
+api.sessions.remove(id /*, jsonApiOptions */)
+
+api.accounts.add({username: 'pat', password: 'secret'/*, [jsonApiOptions] */})
+api.accounts.find(id /*, jsonApiOptions */)
```
Comment (patriciagarcia) [inline api/README.md]: `id` here is the bearer token?

File: api/README.md [hunk 9]
```diff
@@ -0,0 +1,34 @@
...
+})
+```
+
+All methods return promises
+
+```js
+api.sessions.add({username: 'pat', password: 'secret'/*, [jsonApiOptions] */})
+api.sessions.find(id /*, jsonApiOptions */)
+api.sessions.remove(id /*, jsonApiOptions */)
+
```
Comment (patriciagarcia) [inline api/README.md]: `id` here is the bearer token?

File: api/README.md (hunk 8 above)
Comment (patriciagarcia) [inline api/README.md]: `id` here is the same as above?

File: how-it-works.md [hunk 10]
```diff
@@ -0,0 +1,105 @@
...
+```
+
+It's recommended to always use the "id role" to grant permissions to databases,
+as usernames can change.
+
+### Exposed account properties by `hapi-couchdb-account-api`
+
+`hapi-couchdb-account-api`’s REST API will only ever expose the following properties:
+
+1. `username` – read only*
```
Comment (patriciagarcia) [inline how-it-works.md]: why the asterisk in `only*`?

File: how-it-works.md [hunk 11]
```diff
@@ -0,0 +1,105 @@
...
+2. `id` – read only
+3. `password` – write only
+
+Usernames can be changed using the `"requests"` API, for which a custom routine
```
Comment (patriciagarcia) [inline how-it-works.md]: what's this `"requests"` API?

File: tests/integration/utils/json-api-request-headers-errors.js [hunk 12]
```diff
@@ -0,0 +1,57 @@
+var lodash = require('lodash')
```
Comment (patriciagarcia) [inline tests/integration/utils/json-api-request-headers-errors.js]: This file is not used anywhere, was committed by mistake by me.

File: tests/integration/session-test.js [hunk 13]
```diff
@@ -0,0 +1,382 @@
...
+        })
+        .get('/_users/org.couchdb.user:pat-doe')
+        .reply(200, {
+          profile: {
+            fullName: 'pat Doe',
+            email: 'pat@example.com'
+          }
+        })
+
+      var sessionWitProfileResponse = require('./fixtures/session-with-profile-response.json')
```
Comment (patriciagarcia) [inline tests/integration/session-test.js]: `sessionWitProfileResponse` -> `sessionWithProfileResponse`

File: utils/session/create.js [hunk 14]
```diff
@@ -0,0 +1,65 @@
...
+
+    if (!bearerToken) {
+      return callback(Boom.badImplementation())
+    }
+
+    bearerToken = bearerToken.pop()
+
+    // TODO: it's not guratneed that the `id:<accountId>` role is 1st
+    var accountId = body.roles[0].substr(3)
+    var session = {
```
Comment (patriciagarcia) [inline utils/session/create.js]: isn't it important to fix this TODO before merging?

**Comment by patriciagarcia** [comment]
Did a quite review but I am finding difficult not to fall asleep, so I have to leave it here for today ;). 
I left some questions and comments, some of them might now make much sense because of sleepiness.. :sleeping:. In any case, feel free to merge for today and I can do a more exhaustive review tomorrow and let you know if there is anything funky.

File: how-it-works.md (hunk 11 above)
Comment (gr2m) [inline how-it-works.md]: The idea is that for the `/session/account` route, all properties would be read-only. To change a password or a username, the user would use the `/requests` API, for which we can define app-specific handlers. See also https://github.com/hoodiehq/account-rest-api/issues/5

File: api/accounts/add.js (hunk 2 above)
Comment (gr2m) [inline api/accounts/add.js]: I'd say that should be clear as we follow the same pattern for all routes

**Pull request closed without merging**